File: src/main/java/xyz/qincai/manhunt/chat/ChatManager.java
package xyz.qincai.manhunt.chat;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import io.papermc.paper.event.player.AsyncChatEvent;
import xyz.qincai.manhunt.ManhuntNG;
import xyz.qincai.manhunt.game.Match;
import xyz.qincai.manhunt.player.PlayerRole;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

public class ChatManager implements Listener {
    private final ManhuntNG plugin;
    private final Map<UUID, ChatMode> playerChatModes = new ConcurrentHashMap<>();

    public ChatManager(ManhuntNG plugin) {
        this.plugin = plugin;
    }

    public ChatMode getChatMode(UUID uuid) {
        ChatMode explicit = playerChatModes.get(uuid);
        if (explicit != null) return explicit;
        PlayerRole role = plugin.getPlayerManager().getRole(uuid);
        return switch (role) {
            case RUNNER -> ChatMode.GLOBAL;
            case HUNTER -> ChatMode.TEAM;
            case SPECTATOR -> ChatMode.GLOBAL;
        };
    }

    public void setChatMode(UUID uuid, ChatMode mode) {
        PlayerRole role = plugin.getPlayerManager().getRole(uuid);
        if (role == PlayerRole.SPECTATOR) return;
        if (mode == getDefaultMode(role)) {
            playerChatModes.remove(uuid);
        } else {
            playerChatModes.put(uuid, mode);
        }
    }

    private ChatMode getDefaultMode(PlayerRole role) {
        return switch (role) {
            case RUNNER -> ChatMode.GLOBAL;
            case HUNTER -> ChatMode.TEAM;
            case SPECTATOR -> ChatMode.GLOBAL;
        };
    }

    public void resetDefaults() {
        playerChatModes.clear();
    }

    public void sendGlobalMessage(Player sender, String message) {
        Component msg = Component.text()
                .append(plugin.getConfigManager().getMessageComponent("chat.global-prefix"))
                .append(Component.text(sender.getName(), NamedTextColor.WHITE))
                .append(plugin.getConfigManager().getMessageComponent("chat.separator"))
                .append(Component.text(message, NamedTextColor.WHITE))
                .build();
        sendToAllGameParticipants(msg);
    }

    public void sendTeamMessage(Player sender, String message) {
        PlayerRole role = plugin.getPlayerManager().getRole(sender.getUniqueId());
        Component msg = Component.text()
                .append(plugin.getConfigManager().getMessageComponent("chat.team-prefix"))
                .append(Component.text(sender.getName(), NamedTextColor.WHITE))
                .append(plugin.getConfigManager().getMessageComponent("chat.separator"))
                .append(Component.text(message, NamedTextColor.WHITE))
                .build();

        switch (role) {
            case HUNTER -> {
                Match match = plugin.getGameManager().getMatch();
                for (UUID uuid : match.getHunterUuids()) {
                    Player p = Bukkit.getPlayer(uuid);
                    if (p != null) p.sendMessage(msg);
                }
            }
            case RUNNER -> {
                UUID runnerUuid = plugin.getGameManager().getMatch().getRunnerUuid();
                if (runnerUuid != null) {
                    Player p = Bukkit.getPlayer(runnerUuid);
                    if (p != null) p.sendMessage(msg);
                }
            }
            case SPECTATOR ->
                sender.sendMessage(plugin.getConfigManager().getMessageComponent("error.spectator-global-only"));
        }
    }

    private void sendToAllGameParticipants(Component message) {
        Match match = plugin.getGameManager().getMatch();
        for (UUID uuid : match.getHunterUuids()) {
            Player p = Bukkit.getPlayer(uuid);
            if (p != null) p.sendMessage(message);
        }
        for (UUID runnerUuid : match.getRunnerUuids()) {
            Player p = Bukkit.getPlayer(runnerUuid);
            if (p != null) p.sendMessage(message);
        }
        for (UUID uuid : match.getSpectatorUuids()) {
            Player p = Bukkit.getPlayer(uuid);
            if (p != null) p.sendMessage(message);
        }
    }

    @EventHandler
    public void onPlayerChat(AsyncChatEvent event) {
        if (!plugin.getGameManager().isGameActive()) return;

        Player player = event.getPlayer();
        UUID uuid = player.getUniqueId();

        PlayerRole role = plugin.getPlayerManager().getRole(uuid);
        Match match = plugin.getGameManager().getMatch();
        boolean isParticipant = role == PlayerRole.HUNTER
                || role == PlayerRole.RUNNER
                || (role == PlayerRole.SPECTATOR && match.getSpectatorUuids().contains(uuid));

        if (!isParticipant) return;

        event.setCancelled(true);

        String message = PlainTextComponentSerializer.plainText().serialize(event.message());
        ChatMode mode = getChatMode(uuid);
        if (mode == ChatMode.TEAM && isTeamSinglePlayer(uuid)) {
            mode = ChatMode.GLOBAL;
        }

        String prefix = mode == ChatMode.GLOBAL ? "[Global]" : "[Team]";

        if (mode == ChatMode.GLOBAL) {
            sendGlobalMessage(player, message);
        } else {
            sendTeamMessage(player, message);
        }

        if (plugin.getConfigManager().isChatLogToConsole()) {
            plugin.getLogger().info("[Manhunt Chat] " + prefix + " " + player.getName() + ": " + message);
        }
    }

    public boolean isTeamSinglePlayer(UUID uuid) {
        PlayerRole role = plugin.getPlayerManager().getRole(uuid);
        Match match = plugin.getGameManager().getMatch();
        return switch (role) {
            case HUNTER -> match.getHunterUuids().size() <= 1;
            case RUNNER -> true;
            case SPECTATOR -> true;
        };
    }
}

File: pumpkin-plugin/chat_manager.py
from typing import Optional

from chat_mode import ChatMode
from player_role import PlayerRole
from game_manager import GameManager
from player_manager import PlayerManager


class ChatManager:
    def __init__(self, game_manager: GameManager, player_manager: PlayerManager):
        self._game_manager = game_manager
        self._player_manager = player_manager
        self._player_chat_modes: dict[str, ChatMode] = {}

    def get_chat_mode(self, uuid: str) -> ChatMode:
        explicit = self._player_chat_modes.get(uuid)
        if explicit is not None:
            return explicit
        role = self._player_manager.get_role(uuid)
        if role == PlayerRole.HUNTER:
            return ChatMode.TEAM
        return ChatMode.GLOBAL

    def set_chat_mode(self, uuid: str, mode: ChatMode) -> None:
        role = self._player_manager.get_role(uuid)
        if role == PlayerRole.SPECTATOR:
            return
        default = self._get_default_mode(role)
        if mode == default:
            self._player_chat_modes.pop(uuid, None)
        else:
            self._player_chat_modes[uuid] = mode

    def _get_default_mode(self, role: PlayerRole) -> ChatMode:
        if role == PlayerRole.HUNTER:
            return ChatMode.TEAM
        return ChatMode.GLOBAL

    def reset_defaults(self) -> None:
        self._player_chat_modes.clear()

    def is_team_single_player(self, uuid: str) -> bool:
        role = self._player_manager.get_role(uuid)
        match = self._game_manager.match
        if role == PlayerRole.HUNTER:
            return len(match.hunter_uuids) <= 1
        return True

    def send_global_message(self, sender_uuid: str, sender_name: str, message: str) -> str:
        prefix = self._player_manager._game_manager._config_manager.get_message("chat.global-prefix") if hasattr(self._player_manager._game_manager, '_config_manager') else "[Global] "
        return f"{prefix}{sender_name}: {message}"

    def send_team_message(self, sender_uuid: str, sender_name: str, message: str) -> str:
        role = self._player_manager.get_role(sender_uuid)
        match = self._game_manager.match

        if role == PlayerRole.HUNTER:
            recipients = match.hunter_uuids
        elif role == PlayerRole.RUNNER:
            recipients = match.runner_uuids
        else:
            return ""

        return f"[Team] {sender_name}: {message}"

    def get_recipients_for_team(self, sender_uuid: str) -> set[str]:
        role = self._player_manager.get_role(sender_uuid)
        match = self._game_manager.match

        if role == PlayerRole.HUNTER:
            return set(match.hunter_uuids)
        elif role == PlayerRole.RUNNER:
            return set(match.runner_uuids)
        return set()

    def get_all_participants(self) -> set[str]:
        match = self._game_manager.match
        return match.runner_uuids | match.hunter_uuids | match.spectator_uuids
